plugins/hls-qualify-imported-names-plugin/src/Ide/Plugin/QualifyImportedNames.hs

```diff
@@ -0,0 +1,259 @@
+  usedIdentifierName :: !Name,
+  usedIdentifierSpan :: !IdentifierSpan
+}
+
+refMapToUsedIdentifiers :: RefMap a -> [UsedIdentifier]
+refMapToUsedIdentifiers = DList.toList . Map.foldlWithKey' folder DList.empty
+  where
+    folder acc identifier spanIdentifierDetailsPairs =
+      DList.fromList (mapMaybe (uncurry (getUsedIdentifier identifier)) spanIdentifierDetailsPairs) <> acc
+
+    getUsedIdentifier :: Identifier -> Span -> IdentifierDetails a -> Maybe UsedIdentifier
+    getUsedIdentifier identifier span IdentifierDetails {..}
+      | Just identifierSpan <- realSrcSpanToIdentifierSpan span
+      , Right name <- identifier
+      , Use `elem` identInfo = Just $ UsedIdentifier name identifierSpan
+      | otherwise = Nothing
+
+occNameToText :: OccName -> Text
+occNameToText = Text.pack . occNameString
+
+updateColOffset :: Int -> Int -> Int -> Int
+updateColOffset row lineOffset colOffset
+  | row == lineOffset = colOffset
+  | otherwise = 0
+
+usedIdentifiersToTextEdits :: Range -> NameEnv [ImportedBy] -> Text -> [UsedIdentifier] -> [TextEdit]
+usedIdentifiersToTextEdits range nameToImportedByMap sourceText usedIdentifiers
+  | let sortedUsedIdentifiers = sortOn usedIdentifierSpan usedIdentifiers =
+      State.evalState (makeStateComputation sortedUsedIdentifiers) (Text.lines sourceText, 0, 0)
+  where
+    folder :: [TextEdit] -> UsedIdentifier -> State ([Text], Int, Int) [TextEdit]
+    folder prevTextEdits (UsedIdentifier identifierName identifierSpan)
+      | Just importedBys <- lookupNameEnv nameToImportedByMap identifierName
+      , Just (ImportedBy alias _) <- find (isRangeWithinImportedBy range) importedBys
+      , let IdentifierSpan row startCol endCol = identifierSpan
+      , let identifierRange = identifierSpanToRange identifierSpan
+      , let aliasText = Text.pack $ moduleNameString alias
+      , let identifierText = Text.pack $ occNameString $ nameOccName identifierName
+      , let qualifiedIdentifierText = aliasText <> "." <> identifierText = do
+          (sourceTextLines, lineOffset, updateColOffset row lineOffset -> colOffset) <- State.get
+          let lines = List.drop (row - lineOffset) sourceTextLines
+          let (replacementText, remainingLines) =
+                if | line : remainingLines <- lines
+                   , let lineStartingAtIdentifier = Text.drop (startCol - colOffset) line
+                   , Just (c, _) <- Text.uncons lineStartingAtIdentifier
+                   , let isParenthesized = c == '('
+                   , let isBackticked = c == '`'
+                   , let replacementText =
+                           if | isParenthesized -> "(" <> qualifiedIdentifierText <> ")"
+                              | isBackticked -> "`" <> qualifiedIdentifierText <> "`"
+                              | otherwise -> qualifiedIdentifierText ->
+                       (replacementText, lineStartingAtIdentifier : remainingLines)
+                   | otherwise -> (qualifiedIdentifierText, lines)
```

Comment:
Do you check whether the identifier is already qualified?

Reply:
The `qualifiedIdentifierText` comes from a `Name` which contains only a `OccName` that is always unqualified (to my knowledge), so qualifying it should always be safe.

I've added a test to show qualified names are not double qualified.